[DTensor] Support convert StridedShard to shard order and vice versa

zpcore opened the pull request:
We plan to use `StridedShard` to express `shard_order`. This PR adds the function to support the conversion between `StridedShard` and `shard_order`.

I moved some test related function into torch/testing/_internal/common_utils.py. We may only care about **_dtensor_spec.py** and **test_utils.py** in this PR for the review.

### How to convert shard order to StridedShard:
Considering the example:
- placements = $[x_0, x_1, x_2, x_3, x_4]$, all $x_?$ are shard on the same tensor dim.

Let's see how the shard order will impact the split_factor (sf). We loop from right to left in the placements to construct the split_factor by assuming different shard order. Starting from $x_4$, this should be a normal shard.

Then $x_3$. There are two possibilities, $x_3$'s order can be before $x_4$. If so, $x_3$'s sf=1, because $x_3$ is before $x_4$ in the placements. Else $x_3$'s order is after $x_4$, then the $x_3$'s sf should be the mesh dim size of $x_4$, which is $T(x_4)$:
<img width="820" height="431" alt="image" src="https://github.com/user-attachments/assets/f53b4b24-2523-42cc-ad6f-41f3c280db70" />


We can use this method to decide on the split factor for $x_2$, $x_1$ and so on.

### How to convert StridedShard to shard order:
This follows the same method above. We check all possible paths and use the real split_factor to see which path matchs the split_factor. If no such matches, the StridedShard is unable to be converted to shard order.

---
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #167300
* __->__ #166740



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

zpcore committed:
Support convert StridedShard to shard order and vice versa

[ghstack-poisoned]

Skylion007 commented on torch/testing/_internal/common_utils.py:
```diff
@@ -5878,3 +5878,127 @@ def wrapper(self, *args, **kwargs):
+    # Generate all possible sharding placement of tensor with rank
+    # `tensor_rank` over mesh.
+    def _split_list(lst: list, N: int):
+        def compositions(n, k):
```
Isn't this equivalent to:
'''python
def compositions(n: int, k: int):
    # yields lists of length k, positive ints summing to n
    for cuts in combinations(range(1, n), k - 1):
        # add 0 and n as sentinels, then take consecutive differences
        yield [b - a for a, b in pairwise((0, *cuts, n))]
'''
Using itertools

zpcore replied:
Oh, good point to think of the split as placing cuts on the list! Updated with this solution.

zpcore committed:
Update on "[DTensor] Support convert StridedShard to shard order and vice versa"

We plan to use `StridedShard` to express `shard_order`. This PR adds the function to support the conversion between `StridedShard` and `shard_order`.

I moved some test related function into torch/testing/_internal/common_utils.py. We may only care about **_dtensor_spec.py** and **test_utils.py** in this PR for the review.





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Support convert StridedShard to shard order and vice versa"

We plan to use `StridedShard` to express `shard_order`. This PR adds the function to support the conversion between `StridedShard` and `shard_order`.

I moved some test related function into torch/testing/_internal/common_utils.py. We may only care about **_dtensor_spec.py** and **test_utils.py** in this PR for the review.





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

ezyang commented on torch/distributed/tensor/_dtensor_spec.py:
```diff
@@ -127,6 +128,60 @@ def compute_default_shard_order(
+                        cur_sf = cur_placement.split_factor
+                    accumulated_sf = 1
+                    find_order = False
+                    for i in range(len(tensor_dim_order) + 1):
```
... so a cubic algorithm?

zpcore replied:
Yes, let M be mesh dim size, N be tensor, the complexity is O(M^2\*N). I can probably improve it to O(M\*N\*lg(M)), but considering the M, N size are <= 4, it should be fine here.

ezyang commented on torch/distributed/tensor/_dtensor_spec.py:
```diff
@@ -127,6 +128,60 @@ def compute_default_shard_order(
+                cur_placement = placements[mesh_dim]
+                if isinstance(cur_placement, Shard) and cur_placement.dim == tensor_dim:
+                    cur_sf = 1
+                    if isinstance(cur_placement, _StridedShard):
```
Since we also have on the books that we want StridedShard to NOT inherit from Shard anymore, we should try hard to avoid writing code in a way that presupposes this

zpcore replied:
sgtm, fixed

ezyang commented on torch/distributed/tensor/_dtensor_spec.py:
```diff
@@ -127,6 +128,60 @@ def compute_default_shard_order(
+        shard_order = []
+        for tensor_dim in range(max_tensor_dim + 1):
+            tensor_dim_order: list[int] = []
+            for mesh_dim in reversed(range(len(placements))):
```
Why does this go backwards?

zpcore replied:
Added details in the PR description, going from right to left in `placements` makes it easier to compare StridedShard split_factor based on the shard order.

ezyang commented on test/distributed/tensor/test_utils.py:
```diff
@@ -766,6 +774,59 @@ def test_2d_mesh_uneven_strided_shard(self):
+    def world_size(self) -> int:
+        return 32
+
+    @with_comms
```
...if this is under LocalTensorMode why do you need `with_comms`? (cc @dzmitry-huba )

zpcore committed:
Update on "[DTensor] Support convert StridedShard to shard order and vice versa"

We plan to use `StridedShard` to express `shard_order`. This PR adds the function to support the conversion between `StridedShard` and `shard_order`.

I moved some test related function into torch/testing/_internal/common_utils.py. We may only care about **_dtensor_spec.py** and **test_utils.py** in this PR for the review.

### How to convert shard order to StridedShard:
Considering the example:
- placements = $[x_0, x_1, x_2, x_3, x_4]$, all $x_?$ are shard on the same tensor dim.

Let's see how the shard order will impact the split_factor (sf). We loop from right to left in the placements to construct the split_factor by assuming different shard order. Starting from $x_4$, this should be a normal shard.

Then $x_3$. There are two possibilities, $x_3$'s order can be before $x_4$. If so, $x_3$'s sf=1, because $x_3$ is before $x_4$ in the placements. Else $x_3$'s order is after $x_4$, then the $x_3$'s sf should be the mesh dim size of $x_4$, which is $T(x_4)$:
<img width="811" height="434" alt="image" src="https://github.com/user-attachments/assets/8acf6b3e-5f97-448d-990b-a6f3de1d1077" />

We can use this method to decide on the split factor for $x_2$, $x_1$ and so on.

### How to convert StridedShard to shard order:
This follows the same method above. We check all possible paths and use the real split_factor to see which path matchs the split_factor. If no such matches, the StridedShard is unable to be converted to shard order.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore commented on torch/distributed/tensor/_dtensor_spec.py:
```diff
@@ -127,6 +128,60 @@ def compute_default_shard_order(
+    def _maybe_convert_StridedShard_to_shard_order(
+        placements: tuple[Placement, ...], mesh: DeviceMesh
+    ) -> Optional[ShardOrder]:
+        """Try convert _StridedShard to ShardOrder."""
```
Added function description.

zpcore committed:
Update on "[DTensor] Support convert StridedShard to shard order and vice versa"

We plan to use `StridedShard` to express `shard_order`. This PR adds the function to support the conversion between `StridedShard` and `shard_order`.

I moved some test related function into torch/testing/_internal/common_utils.py. We may only care about **_dtensor_spec.py** and **test_utils.py** in this PR for the review.

### How to convert shard order to StridedShard:
Considering the example:
- placements = $[x_0, x_1, x_2, x_3, x_4]$, all $x_?$ are shard on the same tensor dim.

Let's see how the shard order will impact the split_factor (sf). We loop from right to left in the placements to construct the split_factor by assuming different shard order. Starting from $x_4$, this should be a normal shard.

Then $x_3$. There are two possibilities, $x_3$'s order can be before $x_4$. If so, $x_3$'s sf=1, because $x_3$ is before $x_4$ in the placements. Else $x_3$'s order is after $x_4$, then the $x_3$'s sf should be the mesh dim size of $x_4$, which is $T(x_4)$:
<img width="811" height="434" alt="image" src="https://github.com/user-attachments/assets/8acf6b3e-5f97-448d-990b-a6f3de1d1077" />

We can use this method to decide on the split factor for $x_2$, $x_1$ and so on.

### How to convert StridedShard to shard order:
This follows the same method above. We check all possible paths and use the real split_factor to see which path matchs the split_factor. If no such matches, the StridedShard is unable to be converted to shard order.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

wconstab commented on test/distributed/tensor/test_utils.py:
```diff
@@ -766,6 +774,61 @@ def test_2d_mesh_uneven_strided_shard(self):
+            all_shard_order = list(shard_iter)
+            import random
+
+            shard_order_choices = random.sample(
```
i'd rather not have the test be nondeterministic.  perhaps take a snapshot (or seed this) so we are getting some fixed sized problem that is repeatable?

wconstab commented on torch/testing/_internal/common_utils.py:
```diff
@@ -5878,3 +5878,125 @@ def wrapper(self, *args, **kwargs):
+    return tuple(placements)
+
+def generate_shard_orders(mesh, tensor_rank):
+    # Generate all possible sharding placement of tensor with rank
```
this function is pretty hard to follow. i could use a little more intuition for what you're doing in here

zpcore replied:
Luckily this is only for test purpose. Actually we have a test of test (`test_generate_shard_orders`) to test `generate_shard_orders` to show the correctness.

wconstab commented on torch/distributed/tensor/_dtensor_spec.py:
```diff
@@ -127,6 +128,177 @@ def compute_default_shard_order(
+                - Calculate split_factor as the product of mesh sizes for all mesh dimensions
+                  that appear:
+                  1. Earlier in the shard order (lower index in mesh_dims)
+                  2. Later in the placement tuple (higher mesh dimension index)
```
it was't clear if this means 1 and 2 or 1 or 2.  i think you meant and, would be good to call that out explicitly

wconstab commented on torch/distributed/tensor/_dtensor_spec.py:
```diff
@@ -127,6 +128,177 @@ def compute_default_shard_order(
+                        f"found {placements[idx]} in {placements=}."
+                    )
+                mesh_dim = mesh_dims[idx]
+                count = math.prod(mesh.size(i) for i in mesh_dims[:idx] if i > mesh_dim)
```
name this split_factor?

wconstab commented on torch/distributed/tensor/_dtensor_spec.py:
```diff
@@ -127,6 +128,177 @@ def compute_default_shard_order(
+
+          Algorithm:
+              1. If no _StridedShard in placements, return default shard order
+              2. Create a list of mesh_dims_order for each tensor dimension
```
hmm. sounds like you need to know the shard order before you can pass step 2 in the algorithm. i am missing something

zpcore replied:
Yes, empty list, updated the description.

wconstab commented on torch/distributed/tensor/_dtensor_spec.py:
```diff
@@ -127,6 +128,177 @@ def compute_default_shard_order(
+        for mesh_dim in reversed(range(len(placements))):
+            cur_placement = placements[mesh_dim]
+            # _StridedShard may not be a subclass of Shard in the future, so write in this way:
+            if isinstance(cur_placement, Shard | _StridedShard):
```
missing 'else' case
(being defensive, as someone might add more placement types)

zpcore replied:
Added the placement type check.

ezyang reviewed (approved):
Not just the algorithm desc, but the description of correctness would be good! The closed form solution is kind of mind bending to understand.

ezyang reviewed (approved):
Not just the algorithm desc, but the description of correctness would be good! The closed form solution is kind of mind bending to understand.

zpcore committed:
Update on "[DTensor] Support convert StridedShard to shard order and vice versa"

We plan to use `StridedShard` to express `shard_order`. This PR adds the function to support the conversion between `StridedShard` and `shard_order`.

I moved some test related function into torch/testing/_internal/common_utils.py. We may only care about **_dtensor_spec.py** and **test_utils.py** in this PR for the review.

### How to convert shard order to StridedShard:
Considering the example:
- placements = $[x_0, x_1, x_2, x_3, x_4]$, all $x_?$ are shard on the same tensor dim.

Let's see how the shard order will impact the split_factor (sf). We loop from right to left in the placements to construct the split_factor by assuming different shard order. Starting from $x_4$, this should be a normal shard.

Then $x_3$. There are two possibilities, $x_3$'s order can be before $x_4$. If so, $x_3$'s sf=1, because $x_3$ is before $x_4$ in the placements. Else $x_3$'s order is after $x_4$, then the $x_3$'s sf should be the mesh dim size of $x_4$, which is $T(x_4)$:
<img width="820" height="431" alt="image" src="https://github.com/user-attachments/assets/f53b4b24-2523-42cc-ad6f-41f3c280db70" />


We can use this method to decide on the split factor for $x_2$, $x_1$ and so on.

### How to convert StridedShard to shard order:
This follows the same method above. We check all possible paths and use the real split_factor to see which path matchs the split_factor. If no such matches, the StridedShard is unable to be converted to shard order.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Support convert StridedShard to shard order and vice versa"

We plan to use `StridedShard` to express `shard_order`. This PR adds the function to support the conversion between `StridedShard` and `shard_order`.

I moved some test related function into torch/testing/_internal/common_utils.py. We may only care about **_dtensor_spec.py** and **test_utils.py** in this PR for the review.

### How to convert shard order to StridedShard:
Considering the example:
- placements = $[x_0, x_1, x_2, x_3, x_4]$, all $x_?$ are shard on the same tensor dim.

Let's see how the shard order will impact the split_factor (sf). We loop from right to left in the placements to construct the split_factor by assuming different shard order. Starting from $x_4$, this should be a normal shard.

Then $x_3$. There are two possibilities, $x_3$'s order can be before $x_4$. If so, $x_3$'s sf=1, because $x_3$ is before $x_4$ in the placements. Else $x_3$'s order is after $x_4$, then the $x_3$'s sf should be the mesh dim size of $x_4$, which is $T(x_4)$:
<img width="820" height="431" alt="image" src="https://github.com/user-attachments/assets/f53b4b24-2523-42cc-ad6f-41f3c280db70" />


We can use this method to decide on the split factor for $x_2$, $x_1$ and so on.

### How to convert StridedShard to shard order:
This follows the same method above. We check all possible paths and use the real split_factor to see which path matchs the split_factor. If no such matches, the StridedShard is unable to be converted to shard order.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Support convert StridedShard to shard order and vice versa"

We plan to use `StridedShard` to express `shard_order`. This PR adds the function to support the conversion between `StridedShard` and `shard_order`.

I moved some test related function into torch/testing/_internal/common_utils.py. We may only care about **_dtensor_spec.py** and **test_utils.py** in this PR for the review.

### How to convert shard order to StridedShard:
Considering the example:
- placements = $[x_0, x_1, x_2, x_3, x_4]$, all $x_?$ are shard on the same tensor dim.

Let's see how the shard order will impact the split_factor (sf). We loop from right to left in the placements to construct the split_factor by assuming different shard order. Starting from $x_4$, this should be a normal shard.

Then $x_3$. There are two possibilities, $x_3$'s order can be before $x_4$. If so, $x_3$'s sf=1, because $x_3$ is before $x_4$ in the placements. Else $x_3$'s order is after $x_4$, then the $x_3$'s sf should be the mesh dim size of $x_4$, which is $T(x_4)$:
<img width="820" height="431" alt="image" src="https://github.com/user-attachments/assets/f53b4b24-2523-42cc-ad6f-41f3c280db70" />


We can use this method to decide on the split factor for $x_2$, $x_1$ and so on.

### How to convert StridedShard to shard order:
This follows the same method above. We check all possible paths and use the real split_factor to see which path matchs the split_factor. If no such matches, the StridedShard is unable to be converted to shard order.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore commented:
> Not just the algorithm desc, but the description of correctness would be good! The closed form solution is kind of mind bending to understand.

Probably I should draft a proof of correctness at some time, especially when it comes to work together with StridedShard from `view` op.

zpcore committed:
Update on "[DTensor] Support convert StridedShard to shard order and vice versa"

We plan to use `StridedShard` to express `shard_order`. This PR adds the function to support the conversion between `StridedShard` and `shard_order`.

I moved some test related function into torch/testing/_internal/common_utils.py. We may only care about **_dtensor_spec.py** and **test_utils.py** in this PR for the review.

### How to convert shard order to StridedShard:
Considering the example:
- placements = $[x_0, x_1, x_2, x_3, x_4]$, all $x_?$ are shard on the same tensor dim.

Let's see how the shard order will impact the split_factor (sf). We loop from right to left in the placements to construct the split_factor by assuming different shard order. Starting from $x_4$, this should be a normal shard.

Then $x_3$. There are two possibilities, $x_3$'s order can be before $x_4$. If so, $x_3$'s sf=1, because $x_3$ is before $x_4$ in the placements. Else $x_3$'s order is after $x_4$, then the $x_3$'s sf should be the mesh dim size of $x_4$, which is $T(x_4)$:
<img width="820" height="431" alt="image" src="https://github.com/user-attachments/assets/f53b4b24-2523-42cc-ad6f-41f3c280db70" />


We can use this method to decide on the split factor for $x_2$, $x_1$ and so on.

### How to convert StridedShard to shard order:
This follows the same method above. We check all possible paths and use the real split_factor to see which path matchs the split_factor. If no such matches, the StridedShard is unable to be converted to shard order.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Support convert StridedShard to shard order and vice versa"

We plan to use `StridedShard` to express `shard_order`. This PR adds the function to support the conversion between `StridedShard` and `shard_order`.

I moved some test related function into torch/testing/_internal/common_utils.py. We may only care about **_dtensor_spec.py** and **test_utils.py** in this PR for the review.

### How to convert shard order to StridedShard:
Considering the example:
- placements = $[x_0, x_1, x_2, x_3, x_4]$, all $x_?$ are shard on the same tensor dim.

Let's see how the shard order will impact the split_factor (sf). We loop from right to left in the placements to construct the split_factor by assuming different shard order. Starting from $x_4$, this should be a normal shard.

Then $x_3$. There are two possibilities, $x_3$'s order can be before $x_4$. If so, $x_3$'s sf=1, because $x_3$ is before $x_4$ in the placements. Else $x_3$'s order is after $x_4$, then the $x_3$'s sf should be the mesh dim size of $x_4$, which is $T(x_4)$:
<img width="820" height="431" alt="image" src="https://github.com/user-attachments/assets/f53b4b24-2523-42cc-ad6f-41f3c280db70" />


We can use this method to decide on the split factor for $x_2$, $x_1$ and so on.

### How to convert StridedShard to shard order:
This follows the same method above. We check all possible paths and use the real split_factor to see which path matchs the split_factor. If no such matches, the StridedShard is unable to be converted to shard order.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Support convert StridedShard to shard order and vice versa"

We plan to use `StridedShard` to express `shard_order`. This PR adds the function to support the conversion between `StridedShard` and `shard_order`.

I moved some test related function into torch/testing/_internal/common_utils.py. We may only care about **_dtensor_spec.py** and **test_utils.py** in this PR for the review.

### How to convert shard order to StridedShard:
Considering the example:
- placements = $[x_0, x_1, x_2, x_3, x_4]$, all $x_?$ are shard on the same tensor dim.

Let's see how the shard order will impact the split_factor (sf). We loop from right to left in the placements to construct the split_factor by assuming different shard order. Starting from $x_4$, this should be a normal shard.

Then $x_3$. There are two possibilities, $x_3$'s order can be before $x_4$. If so, $x_3$'s sf=1, because $x_3$ is before $x_4$ in the placements. Else $x_3$'s order is after $x_4$, then the $x_3$'s sf should be the mesh dim size of $x_4$, which is $T(x_4)$:
<img width="820" height="431" alt="image" src="https://github.com/user-attachments/assets/f53b4b24-2523-42cc-ad6f-41f3c280db70" />


We can use this method to decide on the split factor for $x_2$, $x_1$ and so on.

### How to convert StridedShard to shard order:
This follows the same method above. We check all possible paths and use the real split_factor to see which path matchs the split_factor. If no such matches, the StridedShard is unable to be converted to shard order.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Support convert StridedShard to shard order and vice versa"

We plan to use `StridedShard` to express `shard_order`. This PR adds the function to support the conversion between `StridedShard` and `shard_order`.

I moved some test related function into torch/testing/_internal/common_utils.py. We may only care about **_dtensor_spec.py** and **test_utils.py** in this PR for the review.

### How to convert shard order to StridedShard:
Considering the example:
- placements = $[x_0, x_1, x_2, x_3, x_4]$, all $x_?$ are shard on the same tensor dim.

Let's see how the shard order will impact the split_factor (sf). We loop from right to left in the placements to construct the split_factor by assuming different shard order. Starting from $x_4$, this should be a normal shard.

Then $x_3$. There are two possibilities, $x_3$'s order can be before $x_4$. If so, $x_3$'s sf=1, because $x_3$ is before $x_4$ in the placements. Else $x_3$'s order is after $x_4$, then the $x_3$'s sf should be the mesh dim size of $x_4$, which is $T(x_4)$:
<img width="820" height="431" alt="image" src="https://github.com/user-attachments/assets/f53b4b24-2523-42cc-ad6f-41f3c280db70" />


We can use this method to decide on the split factor for $x_2$, $x_1$ and so on.

### How to convert StridedShard to shard order:
This follows the same method above. We check all possible paths and use the real split_factor to see which path matchs the split_factor. If no such matches, the StridedShard is unable to be converted to shard order.

---



cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]
